src/System.Linq/src/System/Linq/SelectMany.cs

```diff
@@ -133,5 +123,128 @@ public static partial class Enumerable
             }
         }
+
+        private sealed class SelectManySingleSelectorIterator<TSource, TResult> : Iterator<TResult>, IIListProvider<TResult>
+        {
+            private readonly IEnumerable<TSource> _source;
+            private readonly Func<TSource, IEnumerable<TResult>> _selector;
+            private IEnumerator<TSource> _sourceEnumerator;
+            private IEnumerator<TResult> _subEnumerator;
+
+            internal SelectManySingleSelectorIterator(IEnumerable<TSource> source, Func<TSource, IEnumerable<TResult>> selector)
+            {
+                Debug.Assert(source != null);
+                Debug.Assert(selector != null);
+
+                _source = source;
+                _selector = selector;
+            }
+
+            public override Iterator<TResult> Clone()
+            {
+                return new SelectManySingleSelectorIterator<TSource, TResult>(_source, _selector);
+            }
+
+            public override void Dispose()
+            {
+                if (_subEnumerator != null)
+                {
+                    _subEnumerator.Dispose();
+                    _subEnumerator = null;
+                }
+
+                if (_sourceEnumerator != null)
+                {
+                    _sourceEnumerator.Dispose();
+                    _sourceEnumerator = null;
+                }
+
+                base.Dispose();
+            }
+
+            public int GetCount(bool onlyIfCheap)
+            {
+                if (onlyIfCheap)
+                {
+                    return -1;
+                }
+
+                int count = 0;
+
+                foreach (TSource element in _source)
+                {
+                    checked
+                    {
+                        count += _selector(element).Count();
+                    }
+                }
+
+                return count;
+            }
+
+            public override bool MoveNext()
+            {
+                switch (_state)
+                {
+                    case 1:
+                        // Retrieve the source enumerator.
+                        _sourceEnumerator = _source.GetEnumerator();
+                        _state = 2;
+                        goto case 2;
+                    case 2:
+                        // Take the next element from the source enumerator.
+                        if (!_sourceEnumerator.MoveNext())
+                        {
+                            break;
+                        }
+
+                        TSource element = _sourceEnumerator.Current;
+
+                        // Project it into a sub-collection and get its enumerator.
+                        _subEnumerator = _selector(element).GetEnumerator();
+                        _state = 3;
+                        goto case 3;
+                    case 3:
+                        // Take the next element from the sub-collection and yield.
+                        if (!_subEnumerator.MoveNext())
```

Comment:
It is a common practice in LINQ. See, for example implementation of SelectEnumerableIterator
Since client can stop enumerating at any time, voluntarily or due to a fault, the client needs to be able to dispose the entire internal state of the iterator. Thus we create a composite Dispose.
We could stop just at that, but we often, as a good effort, we dispose on state transitions as well. That is mostly to release resources early and does not need to be hardened for faulting scenarios.